usb/ehci/ehci-hub.c

```diff
@@ -170,23 +233,87 @@ static int ehci_clearPortFeature(usb_dev_t *hub, int port, uint16_t wValue)
 }
 
 
-static int ehci_getDesc(usb_dev_t *hub, char *buf, size_t len)
+static int ehci_getStringDesc(usb_dev_t *hub, int index, char *buf, size_t size)
 {
-	hcd_t *hcd = hub->hcd;
-	usb_hub_desc_t *desc;
+	usb_string_desc_t *desc;
+	size_t len = 0;
+	const char *src;
+	int i;
+
+	switch (index) {
+		case 0:
+			/* LangId string */
+			len = 4;
+			src = ehci_descs.langID;
+			break;
+		case 1:
+			/* Product string */
+			len = strlen(ehci_descs.product) * 2 + 3;
+			src = ehci_descs.product;
+			break;
+		case 2:
+			/* Manufacturer string */
+			len = strlen(ehci_descs.manufacturer) * 2 + 3;
+			src = ehci_descs.manufacturer;
+			break;
+		default:
+			return 0;
+	}
 
-	if (len < 9)
-		return -1;
+	if (size < len)
+		return len;
+
+	desc = (usb_string_desc_t *)buf;
+	desc->bDescriptorType = USB_DESC_STRING;
+	desc->bLength = len;
+
+	if (index == 0) {
+		memcpy(buf, src, len - 2);
+	}
+	else {
+		/* Unicode encode */
+		memset(desc->wData, 0, len - 2);
+
+		for (i = 0; i < strlen(src); i++)
+			desc->wData[i * 2] = src[i];
```

Comment:
Without call `strlen()` in loop, less code and faster.
```suggestion
		for (i = 0; src[i] != '\0'; i++)
			desc->wData[i * 2] = src[i];
```

Reply:
I agree it looks and works better. Thanks.